jax/ops/scatter.py

```diff
@@ -26,33 +26,85 @@
+def _is_advanced_int_indexer(idx):
+  _int = lambda aval: not aval.shape and onp.issubdtype(aval.dtype, onp.integer)
+  try:
+    abstract_idx = core.get_aval(idx)
+  except TypeError:
+    abstract_idx = None
+  out = not (isinstance(abstract_idx, ConcreteArray) and _int(abstract_idx) or
+             isinstance(abstract_idx, ShapedArray) and _int(abstract_idx) or
+             isinstance(idx, slice) or
+             isinstance(idx, tuple) and all(onp.ndim(elt) == 0 for elt in idx))
+  return out and np._is_advanced_int_indexer(idx)
+
 def _scatter_update(x, idx, y, scatter_op):
   """Helper for indexed updates.
 
   Computes the value of x that would result from computing::
     x[idx] op= y
   except in a pure functional way, with no in-place updating.
 
-  Support NumPy-style basic indexing only, i.e., `idx` must be
-  `None`, an integer, a `slice` object, or ellipses, or a tuple of the above.
+  Args:
+    x: ndarray to be updated.
+    idx: None, an integer, a slice, an ellipsis, an ndarray with integer dtype,
+      or a tuple of those indicating the locations of `x` into which to scatter-
+      update the values in `y`.
+    y: values to be scattered.
+    scatter_op: callable, either lax.scatter or lax.scatter_add.
 
-  TODO(phawkins): support advanced indexing.
+  Returns:
+    An ndarray representing an updated `x` after performing the scatter-update.
   """
+  # For more clues on the logic of this implementation, see the code for
+  # jax.numpy._rewriting_take (which has links to NumPy docs).
 
   x = np.asarray(x)
   y = np.asarray(y)
   x_shape = np.shape(x)
   y_shape = np.shape(y)
   y = lax.convert_element_type(y, lax.dtype(x))
 
+  # Check if there's advanced indexing going on, and handle differently based on
+  # whether it is or isn't mixed with basic indexing.
+  if _is_advanced_int_indexer(idx):
+    if np._is_advanced_int_indexer_without_slices(idx):
+      if isinstance(idx, (tuple, list)):
+        if any(onp.shape(e) for e in idx):
+          # At least one sequence element in the index list means broadcasting.
+          idx = np.broadcast_arrays(*idx)
+        else:
+          # The index list is a flat list of integers.
+          idx = [lax.concatenate([lax.reshape(e, (1,)) for e in idx], 0)]
+      else:
+        # The indexer is just a single integer array.
+        idx = [idx]
+
+      stacked_idx = np.concatenate(
+          [np.mod(np.reshape(a, (-1, 1)), np._constant_like(a, x.shape[i]))
+          for i, a in enumerate(idx)], axis=1)
+
+      y = np.broadcast_to(y, idx[0].shape + onp.shape(x)[len(idx):])
+      y = lax.reshape(y, (stacked_idx.shape[0],) + onp.shape(x)[len(idx):])
+
+      dnums = lax.ScatterDimensionNumbers(
+          update_window_dims=tuple(range(1, y.ndim)),
+          inserted_window_dims=tuple(range(len(idx))),
+          scatter_dims_to_operand_dims=tuple(range(len(idx))))
+      return scatter_op(x, stacked_idx, y, dnums)
+    elif np._is_advanced_int_indexer(idx):
```

Comment:
I'm wondering if this case would be handled if you just merged the "sequence of advanced indexes" case into the loop that handles basic indices below? It looks plausibly correct from a quick read of the numpy docs. Not mandatory for this change...

Reply:
I would very much like that to be the case, but my current guess is that it's not going to work. Hopefully I'll be proved wrong, but in any case maybe we can figure it out in a GVC soon.